### ISSUE TITLE
useChannelStateContext doesn't pick up presence updates

### ISSUE THREAD

**Issue by linhewinner** [GetStream/stream-chat-react#1326]
I verified that "user.presence.changed" events are coming through from websocket in Chrome dev tool.

My component that uses useChannelStateContext doesn't get re-rendered when the updates happen. I had to put a timer to re-render my component every minute to pick up the change.

Component ChannelList appears to try to pick up presence updates here:
https://github.com/GetStream/stream-chat-react/blob/2b26d54a2981aa42abd81c8bb743f0a70b4e0df8/src/components/ChannelList/ChannelList.tsx#L283

The presence listener hook, instead of creating new objects, just modify the existing objects:
https://github.com/GetStream/stream-chat-react/blob/2b26d54a2981aa42abd81c8bb743f0a70b4e0df8/src/components/ChannelList/hooks/useUserPresenceChangedListener.ts#L38

Since React won't do deep inspection for object equality, it makes sense that a re-render won't be triggered.

Is there anything we can do better here?
useChannelStateContext gives off the impression that if ANYTHING in the channel state changes, re-render would be automatically triggered and we can write UI code cleanly off of that. That doesn't appear to be the case, and it's very unclear exactly what data changes would trigger re-render and what wouldn't.

**Comment by petyosi** [comment]
Hey @linhewinner,

Based on what you describe, I assume you would like to customize the preview of the channel list so that you can display online/offline users. 

Here's a snippet for a custom preview component that worked for me - notice that I am displaying who' online and who is offline. I modified the example [from our help article](https://getstream.io/chat/docs/sdk/react/custom-code-examples/channel_list_preview/#the-preview-prop-component). 

Let me know if this helps. 

```tsx
const CustomPreview = (props: ChannelPreviewUIComponentProps) => {
  const { channel, setActiveChannel } = props;

  const { channel: activeChannel } = useChatContext();

  const selected = channel?.id === activeChannel?.id;

  const renderMessageText = () => {
    const lastMessageText = channel.state.messages[channel.state.messages.length - 1].text;

    const text = lastMessageText || 'message text';

    return text.length < 60 ? lastMessageText : `${text.slice(0, 70)}...`;
  };

  if (!channel.state.messages.length) return null;

  return (
    <div
      className={selected ? 'channel-preview__container selected' : 'channel-preview__container'}
      onClick={() => setActiveChannel!(channel)}
    >
      <div className='channel-preview__content-wrapper'>
        <div className='channel-preview__content-top'>
          <p className='channel-preview__content-name'>{channel.data?.name || 'Channel'}</p>
          <p className='channel-preview__content-name'>{channel.data?.subtitle}</p>
        </div>
        <p className='channel-preview__content-message'>{renderMessageText()}</p>
        <ul>
          {Object.values(channel.state.members).map((member) => (
            <li>
              {member!.user!.id} - {member!.user!.online ? 'online' : 'offline'}
            </li>
          ))}
        </ul>
      </div>
    </div>
  );
};
```

**Comment by linhewinner** [comment]
This is more of less the code I wrote myself :) and it is precisely how you reproduce the bug!

UserA loads the chat in Chrome. UserB loads the chat in Chrome on a different computer.
UserA sees "UserB is online". Now UserB closes the Chrome tab ....

UserA continues to see "UserB is online" ... (bug: realtime presence update isn't working).
If UserA reload the tab, you'd see "UserB is offline".

When I debugged, the "presence.updated" event is received by UserA's websocket. channel.state.members.online is actually updated to "false", but the React component isn't re-rendered.

Does that make sense?
Can you please reopen this issue?

**Comment by petyosi** [comment]
@linhewinner it makes sense. Investigating this further, the `channelStateContext` is, indeed, not refreshed when online presence changed - the reasons for that being performance. 

Fortunately, there's an easy way to achieve what you're looking for with some additional event listeners. I created this PR to document the process: https://github.com/GetStream/stream-chat-react/pull/1350. Let me know if it matches your requirements.

**Comment by linhewinner** [comment]
@petyosi  Yup! Your code is almost exactly what I wrote in my app to make this work.

You are missing a couple of things though :)
I made a couple of comments.

Before you close this issue, I'd also suggest putting some documentation for useChannelStateContext. The hook gives off the impression that any time anything in the Channel object changes, you'd trigger a re-render. If that's the case, the documentation should set that expectation and preferably even reference to the code example you put up.
